### PR TITLE
Fix zero value metric

### LINKED ISSUE #61: Expression Calculations On 0's Return Nulls Incorrectly
## Description
Welp, the fix we instituted in order to fix the `division by zero` issue is now causing issues in expression metrics that are simple calculations.

### Example
```
select *
from 
{{ metrics.calculate(metric('multi_dimension_expression_metric'), 
    grain='month',
    dimensions=['had_discount','order_country']
    )
}}
```
Here the expression metric is defined as `base_sum_metric` + 1.  But because we override the 0 value with null (to correctly disallow division by zero) we now have issues with calculations.

image

### PR DESCRIPTION
**Pull request by dave-connors-3**
## What is this PR?
This is a:
- [ ] documentation update
- [x] bug fix with no breaking changes
- [ ] new functionality
- [ ] a breaking change

All pull requests from community contributors should target the `main` branch (default).

## Description & motivation
Closes #61 

This initial approach parses expression metrics, checks for a `/` and wraps any divisors in the expression in the appropriate `nullif( {{ dvisor }} , 0)` expression. This approach solves the current issue:

<details>
<summary>Sample Metrics</summary>

```yml
version: 2

metrics:
  
  - name: total_value
    label: Total Value
    model: ref("incremental")
    type: sum
    sql: val

    dimensions:
      - state
      - customer_id
    
    filters:
      - field: state
        operator: '!='
        value: "'connecticut'"

    timestamp: report_date
    time_grains: [day, week, month]


  - name: non_division_exp
    label: Expression ($)
    timestamp: order_date
    time_grains: [day, week, month]
    type: expression
    sql: "{{metric('total_value')}} + 1"
    dimensions:
      - state
      - customer_id

  - name: division_exp_1
    label: Expression ($)
    timestamp: order_date
    time_grains: [day, week, month]
    type: expression
    sql: "1 /{{metric('total_value')}}"
    dimensions:
      - state
      - customer_id

  - name: division_exp_2
    label: Expression ($)
    timestamp: order_date
    time_grains: [day, week, month]
    type: expression
    sql: "1 / {{metric('total_value') }} / 2"
    dimensions:
      - state
      - customer_id

  - name: division_exp_3
    label: Expression ($)
    timestamp: order_date
    time_grains: [day, week, month]
    type: expression
    sql: "{{metric('division_exp_1') }} / 2"
    dimensions:
      - state
      - customer_id


```

</details>

<details>
<summary>Compiled SQL</summary>

```sql
...
, first_join_metrics as (

    select
        date_week
        , coalesce(
            total_value__final.state
            , NULL
        ) as state
        , coalesce(
            total_value__final.customer_id
            , NULL
        ) as customer_id
    
        , total_value as total_value  

    from
        total_value__final 
    )
    
, join_metrics__998 as (

    select 
    
        first_join_metrics.*
        , (1  / nullif(total_value, 0)) as division_exp_1
    
    from first_join_metrics


)
    
, join_metrics__999 as (

    select 
    
        join_metrics__998.*
        , (total_value + 1) as non_division_exp
        , (1  / nullif( total_value , 0) / nullif( 2, 0)) as division_exp_2
        , (division_exp_1  / nullif( 2, 0)) as division_exp_3
    
    from join_metrics__998


)

, joined_metrics as (

    select 
        first_join_metrics.date_week
        , first_join_metrics.state
        , first_join_metrics.customer_id
        , coalesce(first_join_metrics.total_value,0) as total_value
        , coalesce(non_division_exp, 0) as non_division_exp
        , coalesce(division_exp_1, 0) as division_exp_1
        , coalesce(division_exp_2, 0) as division_exp_2
        , coalesce(division_exp_3, 0) as division_exp_3




...


```

Results:

image

My open question here is that this treats _observed_ 0 value results (true zeros from my data) the same way as the filled-in 0s from our spining behavior. Is it appropriate to assume that _any_ zero in the source metric should be evaluated by the corresponding expression metric?

</details>

## Checklist
- [X] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [ ] BigQuery
    - [X] Postgres
    - [ ] Redshift
    - [ ] Snowflake
- [ ] I have updated the README.md (if applicable)
- [X] I have added tests & descriptions to my models (and macros if applicable)
- [ ] I have added an entry to CHANGELOG.md

---
### Tenets to keep in mind 
- A metric value should be consistent everywhere that it is referenced
- We prefer generalized metrics with many dimensions over specific metrics with few dimensions
- It should be easier to use dbt’s metrics than it is to avoid them
- Organization and discoverability are as important as precision
- One-off models built to power metrics are an anti-pattern

### CODE REVIEW

**Comment by callum-mcdata** [comment]
>My open question here is that this treats observed 0 value results (true zeros from my data) the same way as the filled-in 0s from our spining behavior. Is it appropriate to assume that any zero in the source metric should be evaluated by the corresponding expression metric?

I think this is the most important thing for us to clarify on this issue. 

### Areas of Concern
- Expression Metrics: if the expression uses division at all then having `0` values can produce failing metrics 
- Secondary Calculations: without `0` values in previous periods, we are unable to calculate many secondary calculations

What would happen if we completely abandoned the 0 coalesce and used null instead?

Relevant slack thread: https://getdbt.slack.com/archives/C02CCBBBR1D/p1660673393965409

**Comment by callum-mcdata** [comment]
@dave-connors-3 crazy idea - what if we push this out into the metric definition itself? Have some sort of parameter in a config block that determines whether nulls are treated as 0's or as NULLS.

This might be the way to keep metric definition flexible while keeping results consistent.

**Comment by dave-connors-3** [comment]
@callum-mcdata -- here's the current state -- want to call out the way that metric configs are being handled in `get_metric_sql`:

### current state
1. they are added to the `metrics_dictionary`  via the `get_metrics_dictionary` macro, as this was the only place that the metric yml itself gets parsed in the Jinja context (this actually takes place in `develop` and `calculate` and the dict is passed to `get_metric_sql` 
2. they are then moved to a separate `metric_config_dict`, which gets passed to both `build_metric_sql` and `gen_secondary_calculation_cte`. These macros then unpack the configs at the appropriate time to add or remove coalescing as necessary. 

### Considerations
- I chose to do this so that the set of configs gets passed in a consistent way -- depending on the universe of configs we allow users to supply in the future, stands to reason that we have a consistent way of passing them to each macro
- That being said, the unpacking and repassing behavior seems a little redundant -- I was hesitant to pass the entire metric dictionary (with all the core attributes) to _every_ place we might need configs, but I also recognize that the unpacking and repackaging of the configs from the `metric_dictionary` is a bit redundant

@joellabes -- would love your eyes on this too!

File: macros/sql_gen/gen_joined_metrics_cte.sql [hunk 1]
```diff
@@ -139,11 +139,21 @@
     {%- else %}
         join_metrics__{{previous_cte_number}}.*
     {%- endif %}
-    {%- for metric in metric_tree.ordered_expression_set %}
+    {%- for metric in metric_tree.expression_set %}
```
Comment (dave-connors-3) [inline macros/sql_gen/gen_joined_metrics_cte.sql:142]: there was a merge conflict here at some point -- feels like this wasn't supposed to get changed? @callum-mcdata

File: macros/secondary_calculations/perform_secondary_calculation.sql [hunk 2]
```diff
@@ -1,15 +1,15 @@
-{%- macro perform_secondary_calculation(metric_name, grain, dimensions, calc_config) -%}
+{%- macro perform_secondary_calculation(metric_name, grain, dimensions, calc_config, metric_config_dict) -%}
```
Comment (joellabes) [inline macros/secondary_calculations/perform_secondary_calculation.sql:1]: Perhaps `metric_config` for consistency with `calc_config`? I.e. not specifying the datatype of the argument

File: macros/secondary_calculations/secondary_calculation_period_over_period.sql [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{%- macro default__secondary_calculation_period_over_period(metric_name, grain, dimensions, calc_config) -%}
+{%- macro default__secondary_calculation_period_over_period(metric_name, grain, dimensions, calc_config, metric_config_dict) -%}
```
Comment (joellabes) [inline macros/secondary_calculations/secondary_calculation_period_over_period.sql:1]: Same comment about not including the datatype in the metric (or, I guess I could get on board with renaming `calc_config` to `calc_config_dict`, if you think clarifying the datatype is important? I just like consistency)

File: macros/secondary_calculations/secondary_calculation_period_over_period.sql [hunk 4]
```diff
@@ -11,25 +11,40 @@
...
     {-% else -%}
         {% do exceptions.raise_compiler_error("Bad comparison_strategy for period_over_period: " ~ calc_config.comparison_strategy ~ ". calc_config: " ~ calc_config) %}
     {%- endif -%}
 
 {% endmacro %}
 
-{% macro default__metric_comparison_strategy_difference(metric_name, calc_sql) -%}
-    coalesce({{ metric_name }}, 0) - coalesce(
+{% macro default__metric_comparison_strategy_difference(metric_name, calc_sql, metric_config_dict) -%}
+    {%- if metric_config_dict.get("default_value_null", False) %}
```
Comment (joellabes) [inline macros/secondary_calculations/secondary_calculation_period_over_period.sql:26]: I only tangentially saw the discussion about what to call this config, so if it's already been through five rounds of feedback then feel free to ignore. But: 

This doesn't feel intuitive to know what it does. Is it setting default values to null? Is it the value we provide instead in the event of a null? 

I wonder if it should be `treat_null_as_zero` or something (or its opposite, `are_nulls_meaningful`)? 

This is only for the coalesce behaviour in secondary calculations yeah?

File: macros/sql_gen/build_metric_sql.sql [hunk 5]
```diff
@@ -1,5 +1,6 @@
-{%- macro build_metric_sql(metric_dictionary, grain, dimensions, secondary_calculations, start_date, end_date, calendar_tbl, relevant_periods, calendar_dimensions, dimensions_provided) %}
+{%- macro build_metric_sql(metric_dictionary, metric_config, grain, dimensions, secondary_calculations, start_date, end_date, calendar_tbl, relevant_periods, calendar_dimensions, dimensions_provided) %}
```
Comment (joellabes) [inline macros/sql_gen/build_metric_sql.sql:1]: Oh now I'm doubly distressed about the inconsistent naming metric_config vs metric_config_dict

File: macros/sql_gen/build_metric_sql.sql [hunk 6]
```diff
@@ -1,5 +1,6 @@
...
     
+    {%- set default_value_null = metric_config.get("default_value_null") if metric_config else False  -%}
```
Comment (joellabes) [inline macros/sql_gen/build_metric_sql.sql:3]: Do you have to protect against non-boolean inputs here? This isn't something I've ever had to think about because I didn't realise this could be 100% user-defined without any oversight from us. What happens if someone does 

```yml
metrics: 
   name: my_metric
   config: 
     default_value_null: banana
```
?

File: macros/sql_gen/gen_joined_metrics_cte.sql [hunk 7]
```diff
@@ -61,7 +61,7 @@
         ) as {{dim}}
     {%- endfor %}
     {% for metric_name in metric_tree.parent_set %}
-        , nullif({{metric_name}},0) as {{metric_name}}
+        , {{metric_name}} as {{metric_name}}
```
Comment (joellabes) [inline macros/sql_gen/gen_joined_metrics_cte.sql:64]: This is because we're returning all nulls and making them the BI layer's problem, right?

File: macros/sql_gen/gen_joined_metrics_cte.sql [hunk 8]
```diff
@@ -139,11 +139,21 @@
     {%- else %}
         join_metrics__{{previous_cte_number}}.*
     {%- endif %}
-    {%- for metric in metric_tree.ordered_expression_set %}
+    {%- for metric in metric_tree.expression_set %}
         {%- if metric_tree.ordered_expression_set[metric] == cte_number %}
-        ,({{metrics_dictionary[metric].expression | replace(".metric_value","")}}) as {{metrics_dictionary[metric].name}}
+            {%- set metric_definition = metrics_dictionary[metric].expression %}
```
Comment (joellabes) [inline macros/sql_gen/gen_joined_metrics_cte.sql:144]:
```suggestion
            {%- set metric_expression = metrics_dictionary[metric].expression %}
```

I have not updated the rest of the places this impacts 😬

File: macros/sql_gen/gen_joined_metrics_cte.sql [hunk 9]
```diff
@@ -139,11 +139,21 @@
...
+            {%- if "/" in metric_definition -%}
+                {%- set split_division_metric = metric_definition.split('/') -%}
+                {%- set dividend = split_division_metric[0] -%}
+                {%- set divisors = split_division_metric[1:] | list -%}
+                {%- set expression = dividend ~ " / nullif(" ~ divisors | join(", 0) / nullif(") ~ ", 0)" -%}
```
Comment (joellabes) [inline macros/sql_gen/gen_joined_metrics_cte.sql:149]: Oh dear. This scares me. Is there anything we could be doing to require these to be explicitly separated out into numerator and denominator at definition-time, instead of us trying to detect division at execution-time?

I'm also not entirely clear on what it's doing - can you add a comment showing the before and after?

File: macros/sql_gen/gen_joined_metrics_cte.sql [hunk 10]
```diff
@@ -139,11 +139,21 @@
...
+            {%- else -%}
+                {%- set expression = metric_definition -%}
+            {%- endif %}
+        , ({{ expression | replace(".metric_value","") }}) as {{ metrics_dictionary[metric].name }}
```
Comment (joellabes) [inline macros/sql_gen/gen_joined_metrics_cte.sql:153]: Where is `.metric_value` coming from in the expression that it can be replaced with nothing? It looks like this is the expression field from the metric, but that's going to be something like `{{ metric('abc') }} + {{ metric('def') }}`, right? so where is what looks to be a column of a table coming from?

File: macros/sql_gen/gen_joined_metrics_cte.sql [hunk 11]
```diff
@@ -175,10 +185,11 @@
         , {{ dim }}
         {%- endfor %}
         {%- for metric_name in metric_tree.parent_set %}
-        , coalesce({{metric_name}},0) as {{metric_name}}
+        , {{metric_name}}
         {%- endfor %}  
-        {%- for metric in metric_tree.ordered_expression_set%}
-        , {{metric}}
+        {%- for metric in metric_tree.expression_set %}
+        {#- should this me ordered expression set? -#}
```
Comment (joellabes) [inline macros/sql_gen/gen_joined_metrics_cte.sql:191]:
```suggestion
        {#- should this be ordered expression set? -#}
```

Beyond the typo, also a reminder that you need to answer this note to self

**Review by joellabes — CHANGES_REQUESTED** [review]
Most of my nitpicks are around naming and structure, the bones look good! The approach to division is very distressing to me, but I don't have a better answer. I just get nervous when we try to be clever with user-provided content. 

NB: I haven't checked the pytest files at all.

File: macros/secondary_calculations/secondary_calculation_period_over_period.sql (hunk 4 above)
Comment (dave-connors-3) [reply macros/secondary_calculations/secondary_calculation_period_over_period.sql:26]: the only feedback on naming I got here was that `fill_missing_values` was bad, which I wholeheartedly agree with. 

Really not sure what the right way to represent what's happening here is! We are overriding the outputted values for metrics without observations from 0 (default) to null. Can't really imagine a world where you'd want a value here other than null or 0, so I lean towards keeping it a boolean rather than a user supplying some arbitrary value. 

Naming is tough! @callum-mcdata @jtcohen6 what are your thoughts here?

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 7 above)
Comment (dave-connors-3) [reply macros/sql_gen/gen_joined_metrics_cte.sql:64]: this was to make sure we weren't messing up non-division expression metrics (the original motivation here). If i had `metric_1` with a value of zero, and `metric_2` was `metric_1 + 1`, with this nullif I'd get null instead of 1

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 11 above)
Comment (dave-connors-3) [reply macros/sql_gen/gen_joined_metrics_cte.sql:191]: I think this is actually a note to the group! I don't think I know the answer here

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 1 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql:142]: Hmm that is a bit odd. In the end it shouldn't matter because the key values in `ordered_expression_set` will match the values in the list that is `expression_set` but as long as the tests pass I'm fine with your override

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 8 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql:144]: Ah - this is a good callout. I've tended to use metric definition to refer to the entire object that exists in `metrics_dictionary` for a particular metric. So if this is just the single expression then I think renaming is a good call

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 10 above)
Comment (dave-connors-3) [reply macros/sql_gen/gen_joined_metrics_cte.sql:153]: this was here before I got in here tbh! not sure what the goal here is!

Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql:153]: This is a holdover from the original way that we compiled down the metric! In 1.2 the expression of `{{ metric('abc') }} + {{ metric('def') }}` would have compiled to `abc.metric_value + def.metric_value`. We've resolved that in newer versions **but** its one of those things that I've kept around because it doesn't hurt. If I had to put a deprecation date on it, I'd probably go with 1.4 release of core.

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 9 above)
Comment (dave-connors-3) [reply macros/sql_gen/gen_joined_metrics_cte.sql:149]: happy to add a comment here! where would you expect us to do this parsing if not here? in the creation of the `metric_dictionary`?

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 11 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql:191]: Isn't it fun when you write comments and then don't remember what you actually meant? If this was me then I suspect it was me musing on whether I should only be using the ordered expression set - but thats not an issue. As mentioned in above comment, the values in those two dicts are roughly the same and contain the same metrics (ordered contains additional info on metric depth). Expression set is cleaner because we don't have to pull just the key value

File: macros/get_metric_sql.sql [hunk 12]
```diff
@@ -41,6 +41,14 @@ within the final dataset in order to accomplish base + secondary calc functional
 a custom calendar -#}
 {%- set calendar_tbl = ref(var('dbt_metrics_calendar_model', "dbt_metrics_default_calendar")) -%}
 
+{#- Reducing the metric dictionary to just metric configs that are relevant for secondary calcs
+    right now the metric dictionary only gets passed to the build_metric_sql macro, which has config information already
+-#}
+{%- set metric_config_dict = {} -%}
```
Comment (callum-mcdata) [inline macros/get_metric_sql.sql:47]: @dave-connors-3 can you give me the ELI5 why we're creating a new dict here and not just passing along the metric_dictionary to anywhere that needs this info? 

Background: The reason we created the metric dictionary is because of develop. In the original version of the package we could just pass the metric object everywhere and it was easy peasy. With develop we had to come up with an alternative so that was the dictionary - a sort of proxy metric object. I'm curious to the advantages that we gain by creating a new variable for config information instead of having it live with the metric in the same dict

Comment (dave-connors-3) [reply macros/get_metric_sql.sql:47]: I'm not personally convinced there are any benefits 😅 

Configs are kinda funny to me because they in the grey area between a core component of a metric (which would be a top level metric parameter!) and the rubber meets the road of making a metric come to life. 

It felt more correct to pass the "how to I make this metric look" type of objects around instead of the entire "what is this metric" to macros that control a very small piece of the puzzle. I.e. why would `gen_secondary_metrics_cte` need to know about the available grains, or the model the metric relies on? 

That being said, it's 100% redundant to put the config then immediately pull it out of the metric dict, so the solution doesn't even really feel better than the issue.

File: macros/secondary_calculations/secondary_calculation_period_over_period.sql (hunk 4 above)
Comment (callum-mcdata) [reply macros/secondary_calculations/secondary_calculation_period_over_period.sql:26]: I'm a little less bothered on what to name this because I want to ensure that every single option that can be used with metrics is clearly defined in our metrics page. That being said, I agree that an easily understood name is best.

Definitely aligned on it being a boolean value. Replacing nulls with other values would potentially break some loose contracts we have with consumers around the value always being a numeric.

Naming ideas:
- `null_values_treated_as_zero`
- `null_observations_treated_as_zero`

I kind of like something that may be a bit too long but is **reallly** on the nose about what we're doing

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 9 above)
Comment (dave-connors-3) [reply macros/sql_gen/gen_joined_metrics_cte.sql:149]: as for what it's doing, it's parsing the full expression string into chunks, separated by the division sign. The first element in the list is treated as the numerator (dividend) and everything thereafter is a divisor. The outputted expression is then 

`dividend / nullif(divisor_1, 0) / nullif(divisor_2 ,0) ... nullif(divisor_n ,0)` 

just in case folks want to do a whole lot of division

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 1 above)
Comment (dave-connors-3) [reply macros/sql_gen/gen_joined_metrics_cte.sql:142]: I think `expression_set` actually makes more sense here rather than the key val pairs in the ordered set!

File: macros/get_metric_sql.sql (hunk 12 above)
Comment (callum-mcdata) [reply macros/get_metric_sql.sql:47]: I'm not particularly bothered by passing around the entire "what is this metric" object because it mimics the pattern we would do if we didn't need to convert it to a dict. In that world we'd probably reference `metric.calc_config` instead of this big dict of dicts. My advocacy would be to keep it in the metric dict and then pass that.

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 9 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql:149]: @dave-connors-3 without digging into your logic, what would happen if I had `metric a / (metric b / metric c)`?

Comment (dave-connors-3) [reply macros/sql_gen/gen_joined_metrics_cte.sql:149]: nothing good I'll tell ya that much

`metric a / nullif((metric b, 0) / nullif(metric c),0)` aka An Error

File: macros/get_metric_sql.sql (hunk 12 above)
Comment (dave-connors-3) [reply macros/get_metric_sql.sql:47]: i am on board

**Comment by dave-connors-3** [comment]
@callum-mcdata @joellabes I believe I have addressed the comments here w.r.t. naming, passing the args the appropriate way, and validating configs (that last piece open to feedback!)

last things we need here are:
- consensus on the name
- consensus on the logic (callum just added a great callout -- depends on a lot of assumptions on how division will be written)

once we get there, this should be close!

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 9 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: Dang. Could we parse each division statement into a [DIV0](https://docs.snowflake.com/en/sql-reference/functions/div0.html) instead?

So this would be `DIV0(metric a, DIV0(metric b, metric c))`

Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: Looks like bigquery has [safe_divide](https://cloud.google.com/bigquery/docs/reference/standard-sql/mathematical_functions#safe_divide).

Looks like redshift and postgres don't have this native functionality.

Comment (joellabes) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: > what would happen if I had `metric a / (metric b / metric c)`?

Yeah this was what scared me, but I couldn't think of a real-world example and didn't want to scaremonger. 



> Looks like bigquery has [safe_divide](https://cloud.google.com/bigquery/docs/reference/standard-sql/mathematical_functions#safe_divide).

https://github.com/dbt-labs/dbt-utils/issues/679 will bring a cross-db implementation of this, but I doubt it'll handle being given a metric object...

> where would you expect us to do this parsing if not here

I would _prefer_ that we not be parsing at all, and it would be structured yaml: 
```yaml
# models/marts/product/schema.yml
version: 2

models:
 - name: dim_customers
   ...

metrics:
  - name: average_revenue_per_customer
    label: Average Revenue Per Customer
    description: "The average revenue received per customer"

    type: derived
    expression:
       numerator: {{metric('total_revenue')}}
       denominator: {{metric('count_of_customers')}}
    ...
```
I.e. `expression` would be able to be either a string (such as `{{ metric('new_customers') }} + {{ metric('existing_customers') }}`) or a dict with numerator/denominator

@jtcohen6: "We are locking down rc1 of dbt 1.3 next week, no more big changes! 
@joellabes: Hold my beer

Comment (joellabes) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: also I'll just leave these here: 
https://docs.exasol.com/db/latest/sql_references/functions/alphabeticallistfunctions/div.htm
https://www.w3schools.com/sql/func_mysql_div.asp

File: macros/secondary_calculations/secondary_calculation_period_over_period.sql (hunk 4 above)
Comment (joellabes) [reply macros/secondary_calculations/secondary_calculation_period_over_period.sql]: > * `null_values_treated_as_zero`

I'd phrase this in the imperative voice: `treat_null_values_as_zero`. Then it's clearly an instruction to the metrics package.

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 10 above)
Comment (joellabes) [reply macros/sql_gen/gen_joined_metrics_cte.sql:153]: > In 1.2

so does this mean that the currently shipping version of the package, which requires 1.3 or higher, doesn't have this behaviour? In that case couldn't we remove it now instead of waiting for 1.4?

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 9 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: @joellabes I don't think that fixes our issue though? Someone could still put `metric a / (metric b / metric c) in the denominator given that we support arbitrary sql statements. At this point its making me think - can we put some validation to make sure there is only one divide in the entire statement and say we don't support multiple? If a user **really** needed to do the above pattern then they could do
- metric b / metric c == metric d
- metric a/ metric d

**Comment by joellabes** [comment]
Have added my vote for name on that comment (`treat_null_values_as_zero`). 
Consensus on the logic: I've made https://github.com/dbt-labs/dbt-core/issues/5918 to see what the appetite is for turning this into structured data. Otherwise we just wait for it to become a problem.

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 9 above)
Comment (joellabes) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: Yeah I don't think division on division is allowed.

Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: Isn't allowed as in we should prevent that behavior?

Comment (joellabes) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: Sorry that was unclear. I meant we shouldn't include it as a supported paradigm - we won't block you if we see two `/`s in an expression (or an additional `/` in a `denominator` field 🥺), but if it breaks then you're on your own. 

We document it as not recommended, but if you make it work (probably wrapping everything in `nullif`s yourself) then more power to you for sneaking one past the goalie.

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 10 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql:155]: We can punt on this for a different PR.

**Review by callum-mcdata — APPROVED** [review]
this has the stamp of approval! Nice job pushing this over the line, this one was more of a 🐻 than expected

File: macros/sql_gen/gen_joined_metrics_cte.sql (hunk 9 above)
Comment (callum-mcdata) [reply macros/sql_gen/gen_joined_metrics_cte.sql]: For the sake of consistent behavior, we're going to punt on this and @dave-connors-3 will write up an issue with all he learned in the development process. In the future we'll want a more holistic way to address this but for now this helps ensure consistent behavior.

**Comment by callum-mcdata** [comment]
@dave-connors-3 you've done the herculean work here -all yours to hit the big green button.

**Comment by joellabes** [comment]
Love your work, you two!
